nifi-nar-bundles/nifi-stateful-analysis-bundle/nifi-stateful-analysis-processors/src/main/java/org/apache/nifi/processors/stateful/analysis/AttributeRollingWindow.java

```diff
@@ -0,0 +1,325 @@
+        return relationships;
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return properties;
+    }
+
+    @OnScheduled
+    public void onScheduled(final ProcessContext context) throws IOException {
+        timeWindow = context.getProperty(TIME_WINDOW).asTimePeriod(TimeUnit.MILLISECONDS);
+        microBatchTime = context.getProperty(SUB_WINDOW_LENGTH).asTimePeriod(TimeUnit.MILLISECONDS);
+
+        if(microBatchTime == null || microBatchTime == 0) {
+            StateManager stateManager = context.getStateManager();
+            StateMap state = stateManager.getState(SCOPE);
+            HashMap<String, String> tempMap = new HashMap<>();
+            tempMap.putAll(state.toMap());
+            if (!tempMap.containsKey(COUNT_KEY)) {
+                tempMap.put(COUNT_KEY, "0");
+                context.getStateManager().setState(tempMap, SCOPE);
+            }
+        }
+    }
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+
+        try {
+            Long currTime = System.currentTimeMillis();
+            if(microBatchTime == null){
+                noMicroBatch(context, session, flowFile, currTime);
+            } else{
+                microBatch(context, session, flowFile, currTime);
+            }
+
+        } catch (IOException e) {
+            getLogger().error("Failed to save the state for {} due to {}", new Object[] { flowFile, e});
+            session.transfer(flowFile, REL_FAILURE);
+        }
+    }
+
+    private void noMicroBatch(ProcessContext context, ProcessSession session, FlowFile flowFile, Long currTime) throws IOException {
+        final StateManager stateManager = context.getStateManager();
+
+        Map<String, String> state = new HashMap<>(stateManager.getState(SCOPE).toMap());
+        Long count = Long.valueOf(state.get(COUNT_KEY));
+        count ++;
+
+        Set<String> keysToRemove = new HashSet<>();
+
+        for(String key: state.keySet()){
+            if(!key.equals(COUNT_KEY)){
+                Long timeStamp = Long.decode(key);
+
+                if(currTime - timeStamp > timeWindow) {
+                    keysToRemove.add(key);
+                    count --;
+
+                }
+            }
+        }
+        String countString = String.valueOf(count);
+
+        for(String key: keysToRemove){
+            state.remove(key);
+        }
+
+        Double aggregateValue = 0.0D;
+
+        for(Map.Entry<String,String> entry: state.entrySet()){
+            if(!entry.getKey().equals(COUNT_KEY)){
+                aggregateValue += Double.valueOf(entry.getValue());
+            }
+        }
+
+
+        final Double currentFlowFileValue = context.getProperty(VALUE_TO_TRACK).evaluateAttributeExpressions(flowFile).asDouble();
```

Comment:
In case of NumberFormatException, I think that the flow file should be sent to a specific relationship, otherwise it will remain in the input relationship.